### PR TITLE
expression: rewrite builtin function: MUL

### PR DESCRIPTION
**Pull request by hanfei1991**
to https://github.com/pingcap/tidb/issues/3851
to https://github.com/pingcap/tidb/issues/4080

### CODE REVIEW

File: expression/builtin_arithmetic.go [hunk 1]
```diff
@@ -329,6 +334,122 @@ func (s *builtinArithmeticMinusIntSig) evalInt(row []types.Datum) (val int64, is
...
+	sc := s.ctx.GetSessionVars().StmtCtx
+	a, isNull, err := s.args[0].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	b, isNull, err := s.args[1].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	// TODO: How to check a * b overflow ?
```
Comment (lamxTyler) [inline expression/builtin_arithmetic.go]: Is `IsInf` suffice?

Comment (zz-jason) [inline expression/builtin_arithmetic.go]: isinf is sufficient

**Comment by hanfei1991** [comment]
@zz-jason @lamxTyler PTAL

File: expression/builtin_arithmetic.go [hunk 2]
```diff
@@ -392,8 +392,11 @@ func (s *builtinArithmeticMultiplyRealSig) evalReal(row []types.Datum) (float64,
 	if isNull || err != nil {
 		return 0, isNull, errors.Trace(err)
 	}
-	// TODO: How to check a * b overflow ?
-	return a * b, false, nil
+	result := a * b
+	if math.IsInf(result, 1) || math.IsInf(result, -1) {
```
Comment (lamxTyler) [inline expression/builtin_arithmetic.go]: `math.IsInf(result, 0)` is better?

Comment (lamxTyler) [inline expression/builtin_arithmetic.go]: @hanfei1991

**Comment by hanfei1991** [comment]
@lamxTyler PTAL

**Pull request merged**